pkg/assembler/backends/inmem/vulnMetadata.go

```diff
@@ -0,0 +1,302 @@
+// limitations under the License.
+
+package inmem
+
+import (
+	"context"
+	"fmt"
+	"reflect"
+	"strconv"
+	"time"
+
+	"github.com/guacsec/guac/pkg/assembler/graphql/model"
+	"github.com/vektah/gqlparser/v2/gqlerror"
+)
+
+type vulnerabilityMetadataList []*vulnerabilityMetadataLink
+type vulnerabilityMetadataLink struct {
+	id              uint32
+	vulnerabilityID uint32
+	scoreType       model.VulnerabilityScoreType
+	scoreValue      float64
+	timestamp       time.Time
+	origin          string
+	collector       string
+}
+
+func (n *vulnerabilityMetadataLink) ID() uint32 { return n.id }
+
+func (n *vulnerabilityMetadataLink) Neighbors(allowedEdges edgeMap) []uint32 {
+	out := make([]uint32, 0, 1)
+	if allowedEdges[model.EdgeVulnMetadataVulnerability] {
+		out = append(out, n.vulnerabilityID)
+	}
+	return out
+}
+
+func (n *vulnerabilityMetadataLink) BuildModelNode(c *demoClient) (model.Node, error) {
+	return c.buildVulnerabilityMetadata(n, nil, true)
+}
+
+// Ingest CertifyVuln
+func (c *demoClient) IngestVulnerabilityMetadatas(ctx context.Context, vulnerabilities []*model.VulnerabilityInputSpec, vulnerabilityMetadatas []*model.VulnerabilityMetadataInputSpec) ([]string, error) {
+	// TODO (pxp928): move checks to resolver so all backends don't have to implement
+	if len(vulnerabilities) != len(vulnerabilityMetadatas) {
+		return nil, gqlerror.Errorf("uneven vulnerabilities and vulnerabilityMetadatas for ingestion")
+	}
+	var modelVulnMetadataIDList []string
+	for i := range vulnerabilityMetadatas {
+		vulnMetadata, err := c.IngestVulnerabilityMetadata(ctx, *vulnerabilities[i], *vulnerabilityMetadatas[i])
+		if err != nil {
+			return nil, gqlerror.Errorf("IngestVulnerabilityMetadata failed with err: %v", err)
+		}
+		modelVulnMetadataIDList = append(modelVulnMetadataIDList, vulnMetadata)
+	}
+	return modelVulnMetadataIDList, nil
+}
+
+func (c *demoClient) IngestVulnerabilityMetadata(ctx context.Context, vulnerability model.VulnerabilityInputSpec, vulnerabilityMetadata model.VulnerabilityMetadataInputSpec) (string, error) {
+	return c.ingestVulnerabilityMetadata(ctx, vulnerability, vulnerabilityMetadata, true)
+}
+
+func (c *demoClient) ingestVulnerabilityMetadata(ctx context.Context, vulnerability model.VulnerabilityInputSpec, vulnerabilityMetadata model.VulnerabilityMetadataInputSpec, readOnly bool) (string, error) {
+	funcName := "IngestVulnerabilityMetadata"
+	lock(&c.m, readOnly)
+	defer unlock(&c.m, readOnly)
+
+	var vulnerabilityLinks []uint32
+
+	vulnID, err := getVulnerabilityIDFromInput(c, vulnerability)
+	if err != nil {
+		return "", gqlerror.Errorf("%v ::  %s", funcName, err)
+	}
+	foundVulnNode, err := byID[*vulnIDNode](vulnID, c)
+	if err != nil {
+		return "", gqlerror.Errorf("%v ::  %s", funcName, err)
+	}
+	vulnerabilityLinks = foundVulnNode.vulnMetadataLinks
+
+	var searchIDs []uint32 = vulnerabilityLinks
+
+	// Don't insert duplicates
+	duplicate := false
+	collectedVulnMetadataLink := vulnerabilityMetadataLink{}
```

Comment:
Small style nit here for consistency. When creating vars, and setting to something important, use
```
searchIDs := vulnerabilityLinks
```
When creating, and default value is ok, or setting later:
```
var collectedVulnMetadataLink vulnerabilityMetadataLink
```
An exception is where setting to default value explicitly helps with readability
```
duplicate := false
```
This is good, because we want to know it is false unless set to true. `var duplicate bool` is ok as the default value is false but hurts readability.

Reply:
Got it! Fixed